examples/imagenet/ilsvrc12.h

```diff
@@ -0,0 +1,357 @@
+  while (image_list_file >> image_file_name >> label)
+    file_list.push_back(std::make_pair(image_file_name, label));
+  LOG(INFO) << "Total number of test images is " << file_list.size();
+  size_t num_test_images = file_list.size();
+  num_test_images = 500;
+  for (size_t imageid = 0; imageid < num_test_images; imageid++) {
+    string path = input_folder + "/" + file_list[imageid].first;
+    Tensor image = ReadImage(path);
+    label = file_list[imageid].second;
+    Tensor lb(Shape{1}, singa::kInt);
+    lb.CopyDataFromHostPtr<int>(&label, 1);
+    std::vector<Tensor> input;
+    input.push_back(image);
+    input.push_back(lb);
+    string encoded_str = encoder->Encode(input);
+    if (writer == nullptr) {
+      writer = new BinFileWriter();
+      writer->Open(outfile, kCreate);
+    }
+    writer->Write(path, encoded_str);
+  }
+  if (writer != nullptr) {
+    writer->Flush();
+    writer->Close();
+    delete writer;
+    writer = nullptr;
+  }
+  LOG(INFO) << "Write " << num_test_images << " images into " << outfile;
+}
+
+void ILSVRC::ReadMean(string path) {
+  BinFileReader bfreader;
+  string key, value;
+  bfreader.Open(path);
+  bfreader.Read(&key, &value);
+  auto ret = decoder->Decode(value);
+  bfreader.Close();
+  mean = ret[0];
+}
+/// A wrapper method to spawn a thread to execute Train() method.
+std::thread ILSVRC::AsyncLoadData(int flag, string file, size_t read_size,
+                                  Tensor *x, Tensor *y, size_t *n_read) {
+  return std::thread([=]() { LoadData(flag, file, read_size, x, y, n_read); });
+}
+
+size_t ILSVRC::LoadData(int flag, string file, size_t read_size, Tensor *x,
+                        Tensor *y, size_t *n_read) {
+  x->Reshape(Shape{read_size, 3, kCropSize, kCropSize});
+  y->AsType(kInt);
+  y->Reshape(Shape{read_size});
+//  Tensor tmp_x(Shape{read_size, 3, kCropSize, kCropSize});
+//  Tensor tmp_y(Shape{read_size}, kInt);
+  if (file != last_read_file) {
+    if (reader != nullptr) {
+      reader->Close();
+      delete reader;
+      reader = nullptr;
+    }
+    reader = new BinFileReader();
+    reader->Open(file, 100 << 20);
+    last_read_file = file;
+  }
+  // read the same file this time
+  // read to the end of file last time
+  else if (reader == nullptr) {
+    reader = new BinFileReader();
+    reader->Open(file, 100 << 20);
+  }
+  //Timer timer1, timer2, timer3;
+  //float readtime = 0.0f, transtime = 0.0f;
+  vector<string> images;
+  for (size_t i = 0; i < read_size; i++) {
+    // return false when reading to the end of file.
+    string image_path, image;
+    bool ret = reader->Read(&image_path, &image);
+    if (ret == false) {
+      reader->Close();
+      delete reader;
+      reader = nullptr;
+      break;
+    }
+    images.push_back(image);
+  }
+  //auto iot = timer2.Elapsed();
+  int nimg = images.size();
+  //LOG(INFO) << "num of images: " << nimg;
+  *n_read = nimg;
+
+  //Timer omptime;
+  int nProcessors = omp_get_max_threads();
```

Comment:
nthreads<-- set by users
vector<std::thread*> threads;
for (int i = 1; i < nthreads; i++) {
  threads.push_back(new thread(decodeTransform, i, std::ref(images), std::ref(x), std::ref(y));
}
decodeTransform(0, images, x, y);
for(auto th: threads) {
  th->join();
  delete th;
}